presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -586,4 +582,173 @@ public void prepareFinal()
             });
         }
     }
+
+    /**
+     * {@link SpillableFinalOnlyGroupedAccumulator} enables spilling for {@link FinalOnlyGroupedAccumulator}
+     */
+    private static class SpillableFinalOnlyGroupedAccumulator
+            implements GroupedAccumulator
+    {
+        private final FinalOnlyGroupedAccumulator delegate;
+        private final List<Type> spillingTypes;
+
+        private ObjectBigArray<GroupIdPage> rawInputs = new ObjectBigArray<>();
+        private ObjectBigArray<RowBlockBuilder> blockBuilders;
+        private long rawInputsLength;
+
+        public SpillableFinalOnlyGroupedAccumulator(List<Type> types, FinalOnlyGroupedAccumulator delegate)
+        {
+            this.delegate = requireNonNull(delegate, "delegate FinalOnlyGroupedAccumulator is null");
+            this.spillingTypes = requireNonNull(types, "types is null");
+        }
+
+        @Override
+        public long getEstimatedSize()
+        {
+            return delegate.getEstimatedSize() + (rawInputs == null ? 0 : rawInputs.sizeOf()) + (blockBuilders == null ? 0 : blockBuilders.sizeOf());
+        }
+
+        @Override
+        public Type getFinalType()
+        {
+            return delegate.getFinalType();
+        }
+
+        @Override
+        public Type getIntermediateType()
+        {
+            return new ArrayType(RowType.anonymous(spillingTypes));
+        }
+
+        @Override
+        public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+        {
+            checkState(rawInputs != null && blockBuilders == null);
+            rawInputs.ensureCapacity(rawInputsLength);
+            rawInputs.set(rawInputsLength, new GroupIdPage(groupIdsBlock, page));
```

Comment:
Add a `TODO` to denote that there are tricks to avoid storing the whole page in memory. For distincting case, dedup is available by doing compaction.